### ISSUE TITLE
String representation of components is unnecessarily large

### ISSUE THREAD

**Issue by arr28** [SanchoGGP/ggp-base#256]
Was the cause of an OutOfMemoryError on snap-ci for build 143.

Back in my Sgian Dubh days, I sorted out the toString() method of propnet components to (a) be much smaller and (b) if I remember rightly, automatically colour those propositions / components that were outputting true in a given state.

Not sure if we could usefully replicate (b), but (a) should certainly be possible.

**Comment by SteveDraper** [comment]
Thanks - let me know if you need me to do anything, but just porting what you have in Sgian Dubh sounds like a good plan

**Comment by arr28** [comment]
Okay, so there were two issue here.

Firstly, the string representation was the same as the dot representation.  That means that when debugging, you see lots of superfluous rubbish when looking at components.

Second, when dumping the propnet to file, we built the entire string in memory before writing it out.  It is this second problem that's the cause of the intermittent snap-ci failures.

I've fixed both issues.  We might discover that I've gone too far on the first and removed useful information.  In that case, we can add back whatever would actually be useful.
